dashboard/app/models/karel.rb

```diff
@@ -37,20 +37,44 @@ def self.flower_types
   end
 
   # If type is "Karel" return a 3 entry hash with keys 'maze', 'initial_dirt',
-  # and 'final_dirt', the keys map to 2d arrays blockly can render.
-  # final_dirt is always zeroed out until it is removed from Blockly.
+  # and 'raw', the keys map to 2d arrays blockly can render.
   def self.parse_maze(maze_json)
     maze_json = maze_json.to_json if maze_json.is_a? Array
     maze = JSON.parse(maze_json)
-    map, initial_dirt, final_dirt = (0...3).map { Array.new(maze.length) { Array.new(maze[0].length, 0) }}
+    maze.each_with_index do |row, x|
+      row.each_with_index do |cell, y|
+        # optional +/-
+        # followed by the number
+        # optionally followed by R or P to override default flower color
+        # optionally followed by:
+        #   FC to indicate an old-style "fixed" cloud or
+        #   C or Cany to indicate a new "dynamic" cloud
+        match = /^(\+|-)?(\d+)(R|P)?(FC|Cany|C)?$/.match(cell.to_s)
+        unless match
+          raise ArgumentError.new("Cell (#{x},#{y}) with content \"#{cell}\" is invalid")
+        end
+      end
+    end
+
+    # "Karel" covers both Bee and Farmer levels. As of Jan 2016, Bee has
+    # been converted to look only at RawDirt, but Farmer levels and
+    # shared code still use a combination of "Map" and "InitialDirt".
+    # Once all clientside logic has been converted to use just
+    # "RawDirt", this can be vastly simplified.
```

Comment:
Note that my first task for this upcoming sprint will be to do just this. In other words, this placeholder code (along with `generate_raw_dirt` below) will be removed ASAP.